system/lib/pthread/proxying.c

```diff
@@ -123,6 +127,16 @@ void emscripten_proxy_execute_queue(em_proxying_queue* q) {
     executing_system_queue = 1;
   }
 
+  // When the current thread is adding tasks it locks the queue, but we can
+  // potentially try to execute the queue during the add (from
+  // emscripten_yield). This will deadlock the thread, so only try to take the
+  // lock if the current thread is not adding to the queue. We then hope the
+  // queue is executed later when it is unlocked.
+  // XXX: This could leave to starvation if we never process the queue.
+  if (q->active_thread == pthread_self()) {
```

Comment:
How about `if (q->active_thread  && q->active_thread == pthread_self()) {` to avoid calling pthread_self at all in the non-contended case?

Reply:
I wonder if we could combine with the above `executing_system_queue` check?

At least we could use the same technique of using a TLS variable instead of calling pthread_self?   I guess that only works if this issue only applies to the system queue (which I think it does).

Reply:
At least maybe wrap this whole thing in the `is_system_queue` check?

Reply:
+1 to reusing the existing thread-local variable once this is confirmed to be the right direction. It can be hoisted to the top level for use from multiple functions if necessary.